feat(AlertGroup): Introduce Alert Group component

andybraren opened the pull request:
Alert Groups are used to contain and align consecutive alerts, either inline alongside a page's content or in the top-right corner as toast notifications.

This implementation is based on [this visual design](https://redhat.invisionapp.com/share/MFOCVXG3R5S#/screens/323130343):

![323130343](https://user-images.githubusercontent.com/9122899/50176523-a0c00200-02cd-11e9-9d62-64322ec95019.png)

## Implementation notes

The width of a toast group is set by `--pf-c-alert-group--MaxWidth` with an initial value of 600px. Regular alert groups are fullwidth.

Padding is used to space the toast group in the corner and keep it centered in smaller viewports. `pointer-events: none;` allows users to click through that padding (to click nav items, tabs, etc.) and `pointer-events: initial;` is applied to the group’s direct children to make sure alerts themselves are still clickable.

## Questions

- Do we want to enable developers to position toasts in places other than the top-right? Maybe centered toasts are a use case? We’d need to add two more variables for `bottom` and `left`.
- Should toast groups be vertically scrollable? Scrolling wasn't in PF3, but if developers don’t implement grouped notifications I can imagine some becoming impossible to view without clearing the top-most ones. Unfortunately the `box-shadow` around individual alerts makes any scrollbar position look awkward.

Closes #654.

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit 502cc28d89a78dec00bc7ae2297969bb47f46e5c

https://deploy-preview-1136--pf-next.netlify.com

mcoker commented:
> Should toast groups be vertically scrollable?

My 2 cents, from looking at the designs, I don't think we would want a scrollbar. Looks like the list will only show 3 alerts at a time, and if there are more than 3, the 3rd one says to use the notification drawer. So I don't think there will ever be more than 3 there. I suppose that is the UI to use for both toast and inline alert lists? This is also probably a question for @mcarrano and/or kyle/michael.

mcoker commented on src/patternfly/components/AlertGroup/alert-group.scss:
```diff
@@ -0,0 +1,37 @@
+  width: 100%;
+
+  // Spacing between alerts
+  > .pf-c-alert:not(:last-child) {
```
It's lower specificity and less code to use `> * + *` (the [owl selector](https://alistapart.com/article/axiomatic-css-and-lobotomized-owls), with `margin-top`.

andybraren replied:
Oh that's very cool, thank you for sharing that! Would using `margin-top` in this case be an acceptable violation of [our component guideline](https://pf-next.com/guidelines#components)?

> Elements inside a component never use top margins. The first element touches the top of its component.

mcoker replied:
Ah, nope, it's just referring to the first element, it should touch the top of the component.

andybraren replied:
Great, I've made the change.

mcoker commented on src/patternfly/components/AlertGroup/alert-group.scss:
```diff
@@ -0,0 +1,37 @@
+    right: var(--pf-c-alert-group--m-toast--Right);
+    max-width: var(--pf-c-alert-group--m-toast--MaxWidth);
+    padding: var(--pf-c-alert-group--m-toast--PaddingTop) var(--pf-c-alert-group--m-toast--PaddingRight) var(--pf-c-alert-group--m-toast--PaddingBottom) var(--pf-c-alert-group--m-toast--PaddingLeft);
+    pointer-events: none;
```
this is 🔥🔥🔥

andybraren replied:
Your suggestion below was more 🔥 and allowed us to cut 11 lines of CSS!

mcoker commented:
I like the use of `padding` here to keep the list centered, but I think you can write this in a shorter way and also with the `padding`, the list isn't actually `600px` wide. It's 600px - the left/right padding.

What do you think about this instead...

```
    top: var(--pf-c-alert-group--m-toast--PaddingTop);
    right: var(--pf-c-alert-group--m-toast--PaddingRight);
    pointer-events: none;
    width: calc(100% - calc(var(--pf-c-alert-group--m-toast--PaddingRight) * 2));
    max-width: var(--pf-c-alert-group--m-toast--MaxWidth);
```

mcarrano commented:
@andybraren @mcoker There definitely should not be a scrollbar. Theoretically you could stack as may notifications as you want, but we would never advise more than 3. 

I'm curious about the thinking behind making this a component. Is this intended as a general container that I put a toast alert in?  So if I only have a single alert, do I start with a group of one and then add other alerts to it?

mcoker commented on src/patternfly/components/AlertGroup/alert-group.scss:
```diff
@@ -0,0 +1,37 @@
+    padding: var(--pf-c-alert-group--m-toast--PaddingTop) var(--pf-c-alert-group--m-toast--PaddingRight) var(--pf-c-alert-group--m-toast--PaddingBottom) var(--pf-c-alert-group--m-toast--PaddingLeft);
+    pointer-events: none;
+
+    & > * {
```
I think the `&` here is redundant.

andybraren replied:
Thanks to your awesome suggestion above, messing with `pointer-events` isn't needed at all; we can remove this entirely. 👍

mcoker commented on src/patternfly/components/AlertGroup/alert-group.hbs:
```diff
@@ -0,0 +1,6 @@
+<div class="pf-c-alert-group{{#if alert-group--modifier}} {{alert-group--modifier}}{{/if}}"
```
Is this considered a list? I wonder if we should use a `ul`?

andybraren replied:
Yeah probably. Changed and updated the docs.

mcoker commented:
> I'm curious about the thinking behind making this a component. Is this intended as a general container that I put a toast alert in? So if I only have a single alert, do I start with a group of one and then add other alerts to it?

@mcarrano good question. The way it is now, this would be the container for all toast alerts, to position them in the top/right, whether you have one or 30. However a single inline alert can still be displayed on the page inside the alert group component or just on its own outside of the alert group component, and the alert will look the same visually. However if you have more than 1 inline alert, you would want them to be in the alert group to get the vertical spacing.

mcarrano commented:
Thanks for the clarification @mcoker . I think that all makes sense. It might be good to document that somewhere in the description so consumers understand what this is and how to use it to position toast notifications. There is an issue in the backlog for an Alert and Notification demo that should also help clarify this. https://github.com/patternfly/patternfly-next/issues/625

andybraren commented:
I guess the scrollbar question could be re-phrased: Should AlertGroup provide an ugly-yet-functional safety net (scroll functionality) to assist users if/when developers fail to implement their notification system exactly as we've designed?

I can see why we lean toward no, but I also wonder what % of devs will actually implement the collapsing "Max 3" design in the real world given the seemingly more complex JS logic involved. If we can provide solid React/JS code that works out of the box like everything else then I'm comfortable, but I don't know enough about that world to scope out how difficult that will be.

mcoker commented on src/patternfly/components/AlertGroup/examples/alert-group-example.hbs:
```diff
@@ -0,0 +1,50 @@
+{{#> alert-group}}
+  {{#> alert-item}}
+    {{#> alert alert--modifier="pf-m-success" alert--attribute='aria-label="Success Notification"'}}
+      {{#> alert-icon alert-icon--success="true"}}
```
@srambach taught me a cool thing today. If you checkout the [handlebars partials docs](https://handlebarsjs.com/partials.html), you'll see there are basic partials and block partials. Basic is when there is no content in the partial (you're just including a partial), and block is when you include content with the partial. So with....

```
{{#> alert-icon alert-icon--success="true"}}
{{/alert-icon}}
```

You can just write `{{> alert-icon alert-icon--success="true"}}`

\#sarahisagenius

andybraren commented:
@mcarrano I'm okay with that, and this component supports that behavior as-is, it'll just look like this (unscrollable, bottom-most alert cut off until a higher-up alert is dismissed):

<img width="830" alt="2019-01-02 09 59 21" src="https://user-images.githubusercontent.com/9122899/50597358-2e2a5700-0e75-11e9-847a-9b84791026f8.png">

andybraren commented:
After discussing this PR with @gdoyle1 and @mcarrano there are two more updates to make:

1. The Toast example should include a unique "collapsed" third toast as shown below, which is the behavior we're recommending that developers implement. Instead of "Dismiss All" the shortcut/action should be "View all" (lowercase a) and it should open the notification drawer when clicked. This toast should also include a close button.

<img width="576" alt="2019-01-02 16 54 22" src="https://user-images.githubusercontent.com/9122899/50614530-1a034b80-0eaf-11e9-8264-a5d5f2306ee8.png">

2. The developer documentation should be updated to say that every toast needs to include a close button and should never include an action/shortcut except for the unique "collapsed" toast described above.

gdoyle1 commented:
Updated the visuals [here](https://drive.google.com/open?id=1RbOmDbxp-rkJgJ-rklhhIyppHSbMC3nl) to match the implementation as well @andybraren !

andybraren commented:
I've updated this PR to address the above, but Toast Alerts won't quite match the visual design until the following issues are addressed:

1. The visual design for the third toast uses an "Other" alert type with a non-standard grey color. This wasn't implemented in the Alert component, so I'm using a blue Info type instead. Are we okay with this, or should we go back and add this new alert type to the Alert component? @mcarrano @gdoyle1

<img width="580" alt="2019-01-03 12 00 12" src="https://user-images.githubusercontent.com/9122899/50650667-23dc8b80-0f4f-11e9-8c25-5aa1229fb943.png">

<img width="623" alt="2019-01-03 11 59 00" src="https://user-images.githubusercontent.com/9122899/50650639-09a2ad80-0f4f-11e9-8282-f88dd741a32c.png">

2. The spacing between the close button and secondary action button ("View all" in this case) doesn't match the visual design right now. The Alert component needs to be updated to fix that spacing. I've created a new issue #1172.

3. We might want to consider re-structuring the `.pf-c-alert__action` element to align with other components as described in #774. Right now we squeeze the close button, primary action button, and secondary action button ("View all") within this one div, which seems to unnecessarily constrain the `.pf-c-alert__body` and its longer text (somewhat shown above, just imagine if "View all" were longer in another language).

4. Reiterating #1099, the Alert component should have an example of an alert with a close button using the appropriate `aria-label`. @jgiardino the Button component's example use `aria-label="Remove"`, but in the context of inline alerts and toasts do you think "Dismiss" would be a better word?

gdoyle1 commented:
@andybraren I don't really have a strong opinion on using the blue info type instead of the grey one designed - the intent of designating that alert with a distinct color was because it only applies to a single use case. Although, I do think that use case would generally fall under the category of "info" alerts. @mcarrano Do you think this falls under "general informational messages"?

mcarrano commented:
@andybraren @gdoyle1 a couple of comments...

I think that we should stick with the blue info color for this alert as we are not using that grey color anywhere else for an alert. @kybaker , any problem with that? Not sure why gray was used in the original visual.

In this case, I'm questioning whether both 'View all' and 'X' should both be there. I know we said the user sure always be able to dismiss a toast notification, but I see this as a special case. The third toast should disappear on it's own when there are 3 or less actual notifications either because the user cleared them or because they expired. Does that make sense?

gdoyle1 commented:
@mcarrano I get your point with removing the 'x' on the last toast, but would only say that if it does not have an 'x' then it wouldn't follow the documentation we have on toasts. I don't really see a harm in leaving the 'x' there in case the user wants to immediately dismiss that toast before the other two on top of it. Otherwise, if they were to just 'x' out one on top (or one were to disappear on top), that last toast would disappear automatically (I think that was what you were saying).

andybraren commented:
@mcarrano @gdoyle1 As an impatient user I wouldn't enjoy having to either wait 8 seconds or click more than 3 X's to make all toasts go away. I'd vote for always having an X, even on the special third toast.

And to clarify the behavior of that special third toast, looking at this image for example:

<img width="662" alt="2019-01-03 15 37 24" src="https://user-images.githubusercontent.com/9122899/50660413-affdab80-0f6d-11e9-8ba2-967cd9cc393f.png">

Are we saying that dismissing/closing the second red toast would make the third blue toast dissolve and reveal the remaining two (older) notifications? I wonder if triaging 4+ notifications two at a time like that will feel better than keeping the third toast static as just a heads-up that there are # more alerts in the notification drawer. I also wonder what the animations should be if/when the special toast goes poof.

mcarrano commented:
@andybraren and @gdoyle1 these are all good points.  I just finished a conversation with @matthewcarleton about all of this and we agreed that it may be time to rethink the design of this a bit. There is clearly more complexity to the toast aggregation than initially considered. At the same time we want to avoid adding more complexity to the Alert component itself.

I am going to set up a meeting so we can talk through this live and consider alternatives for keeping this moving forward.

jgiardino commented:
> Reiterating #1099, the Alert component should have an example of an alert with a close button using the appropriate aria-label. @jgiardino the Button component's example use aria-label="Remove", but in the context of inline alerts and toasts do you think "Dismiss" would be a better word?

"Dismiss" sounds like the right action word for the Alerts in a group. 

Also, in similar patterns like ChipGroup and DataList where items in a list have the same action, we're also labeling the action with the main text of that item. This is helpful for providing context for buttons that otherwise would have identical labels. 

For example, the remove button in a chip has the following attributes:
```
  <button class="pf-c-button pf-m-plain" aria-labelledby="remove_chip_one chip_one" aria-label="Remove" id="remove_chip_one">
    <i class="fas fa-times-circle" aria-hidden="true"></i>
  </button>
```
If the Chip label is "Puppies", the button is announced as:
> Remove Puppies button

In this case, the dismiss button should reference the heading element of the Alert:
```
    <h4 class="pf-c-alert__title" id="alert_one_title">
      <span class="pf-u-sr-only">
       				Success: 
     </span> Files are uploaded
    </h4>
```
And the combination of the Dismiss label and the heading text would be announced as: 
> Dismiss Success: Files are uploaded button


But I'd love some suggestions on refinements to this. Like should we include the word "alert" in the hidden text (i.e. "Success alert:"), should we drop the heading text and only reference the type of alert (e.g. "Dismiss Success alert: button") and rely more on context for clarification to avoid potentially really long buttons labels? Given the number of alerts that could display and the likeliness that they would be removed automatically, I don't imagine there being a large number of any one type of alert anyway. 

I'm cc'ing @stacymcauliffe in case she has input, too.

stacymcauliffe commented:
@jgiardino - So, in order, the user would hear something like -

"Success: Files are uploaded" (plus maybe something that indicates to them that this is an alert)

and then when they navigate to the next element, 

"Dismiss Success Files are uploaded button"
 
or

"Dismiss success alert button"...

I *think* it would be enough to read "Dismiss success alert button" but it's hard to say definitively without understanding how easy it is for context to be lost/forgotten. I would imagine they'd be hearing the information contained in the alert message and then immediately either clicking a CTA/action button in the alert or simply dismissing.

jgiardino commented:
@stacymcauliffe 

Yes. There's also the case where they're only scrolling through all the buttons on the page. So they could hear:

"Dismiss Success alert button"
"Dismiss Success alert button"
"Dismiss Warning alert button"

If they're navigating by button, they could shift focus to the previous/next element next to that button to get a sense of context. But they would need to somehow know when they're on a button whether that alert text that comes before it or after it is the alert they're dismissing, which could be challenging. 

I'm still really on the fence with this, but for that reason I slightly lean toward the more verbose, obvious method of labeling.  

When you say
> (plus maybe something that indicates to them that this is an alert)
That isn't there currently, but that would be good. I noted this in the description for #854 but included a question for you in that issue.

mcarrano commented:
Looks good @andybraren . That's for making changes. But just a couple of questions/comments...

* The documentation says this, "Alerts within the group should not include any actions, links, or buttons except for a close button that dismisses the alert." But the example violates it by including an alert with a button. Can we remove this and show all the alerts with a Close button, only? Also, links and actions are allowed as long as they are inline with the text.

* I'm curious what drives the width of the Alert Group. Seems like this should be a fixed width.

jgiardino commented:
@mcarrano Are you looking at the first example? If I have more than one static alert on a page, would those be included in an AlertGroup? And can static alerts include actions?

In the example "Toast Alert Group", I know we discussed suggesting that a maximum number of alerts display would display. But that behavior would be dependent on the developer implementing that. In the case where a developer doesn't implement that behavior and is just popping off a long list of alerts, should we handle that case and allow the alerts to scroll?

andybraren commented:
@mcarrano @jgiardino This AlertGroup component is weird in that it handles and provides examples for both "Inline Alerts" (can contain actions, 4+ can be displayed wherever the developer wants) and "Toast Alerts" (max of 3 visible, links/buttons within the description paragraph, but no action buttons other than close).

@mcarrano that documentation text should be gone. I just added another commit to re-add that "Every alert within a Toast Alert group must include a close button to dismiss the alert." The first "Inline Alert Group" example is full width, the second full-page "Toast Alert Group" example has a restricted width and is placed in the top-right corner.

@jgiardino From the docs: "`.pf-c-alert-group` is optional when only one alert is needed. It becomes required when more than one alert is used in a list." The component basically just adds consistent spacing between consecutive alerts so that they're not directly laying on top of one another. Regarding scrolling, [I agree](https://github.com/patternfly/patternfly-next/pull/1136#issuecomment-449488309).

mcarrano commented:
@andybraren @jgiardino My bad. I misread this and did not see that there were two variants of the alert group. One for inline and another for toast. Do we need to use an alert group to display inline notifications? I thought this was only needed for the toast alerts. The reason I say that is that be definition, an inline alert is just inserted in the page content, it does not float on top like a toast. @matthewcarleton can you confirm this. 

The Toast Alert Group example looks good now.

I don't want to add a scroll bar. For now, let's just let the toast alerts stack. This is the "overflow" handing that I spoke about during the meeting that needs further design. There is a new issue I opened for that here: https://github.com/patternfly/patternfly-next/issues/1181

andybraren commented:
@mcarrano 

> Do we need to use an alert group to display inline notifications?

Only if there's more than 1, discussed [here](https://github.com/patternfly/patternfly-next/pull/1136#issuecomment-449485127). A single inline alert does not need to be in an Alert Group because it will look the same regardless. As soon as a second adjacent inline alert is needed they should both be inside of an Alert Group to ensure that they have the proper spacing between them.

In my opinion, a forward-thinking developer would use an Inline Alert Group to begin with because there's a good chance a second alert could come along later. But we're not forcing their hand.

> I thought this was only needed for the toast alerts.

Alert Groups are inline by default, born out of the need to space multiple inline alerts properly. An Inline Alert Group can be transformed into an identically-spaced but top-right-aligned Toast Alert Group using the `.pf-m-toast` modifier. If the toast area's design changes significantly I can imagine it might need to become its own component, but for now it was similar enough to Inline Alert Groups that achieving that layout with just a modifier for top-right positioning made the most sense.

andybraren commented:
@jgiardino @stacymcauliffe I’ve updated this PR to include `aria-labelledby=“alert_one_title” on every close button with a corresponding `id=“alert_one_title` on the alert’s `<h4>` heading to achieve the screen reader behavior discussed above and in #854. I think the accessibility documentation for these close buttons probably belongs in the Alert component. I've also added accessibility documentation saying that “notifications” should refer to themselves as such in the `.pf-u-sr-only` text.

If this looks good from an accessibility and documentation perspective, I believe this is ready to merge.

jgiardino commented:
Thanks for making these updates! There were a couple of things I noticed:

- Are we referring to these as "Notifications" or "Alerts"? Maybe that's something that was already discussed, but just wanted to double-check. When chatting with @stacymcauliffe, we settled on using this label for both dynamic and static alerts. I'm not sure if one works better for both, so I'll defer to others on this on.
- There are a couple of tweaks needed for the attributes on the Dismiss button. 
  - This is what it is currently, which would be announced as "Success notification: File was uploaded button" (assuming the heading is "File was uploaded"):
   `<button class="pf-c-button pf-m-plain" aria-label="Dismiss" aria-labelledby="notification_three_title">`
  - This is what it should be, which would be announced as "Dismiss Success notification: File was uploaded button":
    `<button class="pf-c-button pf-m-plain" aria-label="Dismiss" aria-labelledby="notification_three_button notification_three_title" id="notification_three_button>`
    - The aria-labelledby attribute will override the aria-label value, and the only way to announce the aria-label text string that is by referencing the button id in the aria-labelledby value

mcoker commented on src/patternfly/components/AlertGroup/docs/toast-group.md:
```diff
@@ -0,0 +1,17 @@
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-u-sr-only` | `.pf-c-alert__title <span>` | Content that is visually hidden but accessible to assistive technologies. This should state the type of notification and refer to itself as a notification rather than an alert.  ** Required**|
```
Can you rebase, then update any instance of `.pf-u-sr-only` in this PR to `.pf-screen-reader`? We're trying to get all of the utility classes out of core.

andybraren replied:
Fixed in latest commit, thanks for pointing that out.

mcoker commented on src/patternfly/components/AlertGroup/examples/alert-group-example.hbs:
```diff
@@ -0,0 +1,54 @@
+      {{/alert-icon}}
+        {{#> alert-body alert-body--attribute='role="alert"'}}
+          {{#> alert-title}}
+            {{#> accessibility accessibility--type="sr-only"}}
```
You can use `{{#> screen-reader}}` instead of `{{#> accessibility}}` now. It's in the same dir as `accessibility.hbs` if you want to look at it. It's the same as `accessibility.hbs` except you don't have to pass `sr-only` to it.

andybraren replied:
Switched to `{{#> screen-reader}}` and aligned the handlebars code more generally with the most recent updates to the Alert component.

andybraren commented:
I’ve updated this PR to:

1. Rebase
2. Switch to `.pf-c-screen-reader` per @mcoker’s review
3. Fix the accessibility issues that [Jenn noted](https://github.com/patternfly/patternfly-next/pull/1136#issuecomment-452880939) (I didn’t know `aria-labelledby` could be concatenated like that, cool, thank you Jenn!)
4. Add `role="alert"` to toast alerts with updated accessibility docs.
5. Remove redundant accessibility documentation now that the Alert component includes examples with close buttons (as of PR #1208)

I’ve also created a new issue #1126 to propose some small updates we should consider making to the Alert component.

patternfly-build commented:
:tada: This PR is included in version 1.0.124 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly-next)
- [GitHub release](https://github.com/patternfly/patternfly-next/releases/tag/v1.0.124)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
